File: Week08/Mod1a-maxflow-ek.cpp
// This code is adapted from:
// https://github.com/stevenhalim/cpbook-code/blob/master/ch8/maxflow.cpp
// This solves Internet Bandwidth - UVA 820
// https://onlinejudge.org/index.php?option=com_onlinejudge&Itemid=8&page=show_problem&category=0&problem=761

#include <bits/stdc++.h>
using namespace std;

typedef long long ll;
typedef tuple<int, ll, ll> edge;
typedef vector<int> vi;
typedef pair<int, int> ii;

const ll INF = 1e18;                             // large enough

class max_flow {
private:
  int V;
  vector<edge> EL;
  vector<vi> AL;
  vi d, last;
  vector<ii> p;

  bool BFS(int s, int t) {                       // find augmenting path
    
    d.assign(V, -1);                             // initialize distance array
    d[s] = 0;                                    // s is the source, so d[s] = 0
    
    queue<int> q({s});                           // load s on the queue 
    
    p.assign(V, {-1, -1});                       // record BFS sp tree

    while (!q.empty()) {
    
      int u = q.front(); q.pop();                // current vertex to be processed

      if (u == t) break;                         // stop as sink t reached
     
      for (auto &idx : AL[u]) {                  // explore neighbors of u
        auto &[v, cap, flow] = EL[idx];          // stored in EL[idx]
        
        if ((cap-flow > 0) && (d[v] == -1)){     // positive residual edge + unexplored
            
            d[v] = d[u]+1;                       // set distance
            q.push(v);                           // push on queue
            p[v] = {u, idx};                     // set parent information
        
        }
      }
    }
    
    return d[t] != -1;                           // has an augmenting path
  
  }

  ll send_one_flow(int s, int t, ll f = INF) {   // send one flow from s->t
    
    if (s == t) return f;                        // bottleneck edge f found
    
    auto &[u, idx] = p[t];    
    auto &cap = get<1>(EL[idx]), &flow = get<2>(EL[idx]);
    
    ll pushed = send_one_flow(s, u, min(f, cap-flow));    
    flow += pushed;
    
    auto &rflow = get<2>(EL[idx^1]);             // back edge
    rflow -= pushed;                             // back flow
    
    return pushed;
    
  }

public:
  max_flow(int initialV) : V(initialV) {
    EL.clear();
    AL.assign(V, vi());
  }

  // if you are adding a bidirectional edge u<->v with weight w into your
  // flow graph, set directed = false (default value is directed = true)
  void add_edge(int u, int v, ll w, bool directed = true) {
    if (u == v) return;                          // safeguard: no self loop
    
    EL.emplace_back(v, w, 0);                    // u->v, cap w, flow 0
    AL[u].push_back(EL.size()-1);                // remember this index
    
    EL.emplace_back(u, directed ? 0 : w, 0);     // back edge
    AL[v].push_back(EL.size()-1);                // remember this index
  
  }

  ll edmonds_karp(int s, int t) {
    
    ll maxf = 0;                                 // maxf stands for max_flow
    
    while (BFS(s, t)) {                          // an O(V*E^2) algorithm
    
      ll f = send_one_flow(s, t);                // find and send 1 flow f
      if (f == 0) break;                         // if f == 0, stop
      maxf += f;                                 // if f > 0, add to maxf
    
    }
    
    return maxf;
  }

};

int main() {

  #ifndef ONLINE_JUDGE 
      freopen("in.txt", "r", stdin);
      freopen("out.txt", "w", stdout);
  #endif
  
  std::ios::sync_with_stdio(false);

  int c = 0; 

  while(true){
    int n; 
    cin >> n;
   
    if(not n) break; 

    int s, t, m;
    cin >> s >> t >> m; 

    max_flow maxf(n+1);

    for(int i = 0; i < m; i++){
      
      int u, v, cap; 
      cin >> u >> v >> cap; 
      maxf.add_edge(u,v,cap,false); 

    }

    c++;  
    cout << "Network " << c << endl << "The bandwidth is " << maxf.edmonds_karp(s,t) << "." << endl << endl; 
  }
  return 0;

}

File: Week08/Mod1a-maxflow-ek.py
INF = 1e18

class MaxFlow():
    EL = []
    AL = []
    d = []
    p = []
    V = 0

    def __init__(self, initialV):
        self.V = initialV
        self.EL = []
        self.AL = [[] for i in range(self.V)]


    def BFS(self, s, t):
        self.d = [-1 for i in range(self.V)]
        self.d[s] = 0

        q = [s]

        self.p = [[-1, -1] for i in range(self.V)]

        while(len(q) != 0):
            u = q.pop(0)

            if u == t:
                break

            for idx in self.AL[u]:
                v, cap, flow = self.EL[idx]

                if cap-flow>0 and self.d[v] == -1:
                    self.d[v] = self.d[u] + 1
                    q.append(v)
                    self.p[v] = [u, idx]

        return self.d[t] != -1

    def send_one_flow(self, s, t, f=INF):

        if s==t:
            return f

        u, idx = self.p[t]

        cap, flow = self.EL[idx][1], self.EL[idx][2]

        pushed = self.send_one_flow(s, u, min(f, cap-flow))
        self.EL[idx][2] += pushed

        self.EL[idx^1][2] -= pushed

        return pushed

    def add_edge(self, u, v, w, directed=True):
        if u==v:
            return
        
        self.EL.append([v, w, 0])
        self.AL[u].append(len(self.EL) - 1)

        self.EL.append([u, 0 if directed else w, 0])
        self.AL[v].append(len(self.EL) - 1)

    def edmonds_karp(self, s, t):
        maxf = 0

        while self.BFS(s, t):
            f = self.send_one_flow(s, t)
            if f == 0:
                break
            maxf += f

        return maxf

c = 0
while True:
    n = int(input())
    if not n:
        break

    s, t, m = map(int, input().split())
    maxf = MaxFlow(n+1)

    for i in range(m):
        u, v, cap = map(int, input().split())
        maxf.add_edge(u, v, cap, False)

    c += 1
    print("Network "+str(c)+"\nThe bandwidth is "+str(maxf.edmonds_karp(s, t))+".\n")
